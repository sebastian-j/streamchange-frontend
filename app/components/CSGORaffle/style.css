.dialog-root {
    display: flex;
    align-items: center;
    justify-content: center;
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
}

.dialog-backdrop {
    background-color: rgba(0, 0, 0, 0.5);
    border: none;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    height: 100%;
    width: 100%;
    outline: none;
}

.raffle-dialog {
    background: linear-gradient(to right, #0f161d 0%, #131b24 51%, #0f161d 100%);
    padding: 0 30px;
    height: 300px;
    width: 800px;
}

.roller-box {
    margin-top: 50px;
    height: 160px;
    background: #000;
    border: 1px solid #616161;
    overflow: hidden;
    position: relative;
    padding: 0;
}

.roller-movable {
    position: absolute;
    left: 0;
    transition-duration: 7s;
    transition-timing-function: cubic-bezier(0.18, 0.17, 0.02, 1);
}

.roller-cell {
    position: relative;
    display: block;
    height: 100%;
    text-align: center;
    color: white;
    line-height: 155px;
    font-size: 0.8em;
    font-weight: bold;
    font-family: sans-serif;
    width: 148px;
}

.roller-cell img {
    height: 100%;
    width: 100%;
}

.roller-label {
    background-color: rgba(0, 0, 0, 0.5);
    display: block;
    height: 40px;
    line-height: 40px;
    position: absolute;
    bottom: 0;
    width: 100%;
    left: 0;
}

.roller-needle {
    width: 2px;
    height: 198px;
    top: 1px;
    left: 50%;
    position: absolute;
    background: #FFCE0A;
    opacity: 0.6;
    z-index: 2;
}

.raffle-winner {
    color: white;
    display: block;
    font-size: 2rem;
    opacity: 0;
    margin-top: 20px;
    text-align: center;
    animation: show-winner 0.5s linear 7.1s;
    animation-fill-mode: forwards;
}
.raffle-winner:after {
    opacity: 1;
}

@keyframes show-winner {
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
}