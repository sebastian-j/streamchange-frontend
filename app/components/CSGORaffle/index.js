import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import db from '../YoutubeWorker/db';
import './style.css';

const CSGORaffle = props => {
  const [users, setUsers] = useState([]);
  const [scrollSize, setScrollSize] = useState(0);
  const [winner, setWinner] = useState(null);
  const [timer, setTimer] = useState(null);

  const closeImmediately = () => {
    props.onClose();
    clearTimeout(timer);
  };

  useEffect(() => {
    db.table('users')
      .filter(user => user.isEligible === true)
      .toArray()
      .then(items => {
        const shuffled = [];
        for (let i = 0; i < 50; i += 1) {
          shuffled.push(items[Math.floor(Math.random() * items.length)]);
        }
        const winnerIndex = Math.floor(Math.random() * 10) + 35;
        const scroll = -(
          winnerIndex * 150 +
          Math.floor(Math.random() * 65) -
          290
        );
        setUsers(shuffled);
        setScrollSize(scroll);
        setWinner(shuffled[winnerIndex]);
        setTimer(
          setTimeout(() => {
            props.onWin(shuffled[winnerIndex].id);
          }, (props.duration + 1) * 1000),
        );
      });
  }, []);

  return (
    <div className="dialog-root">
      <div className="dialog-backdrop" onClick={closeImmediately} />
      <div className="raffle-dialog">
        <div className="roller-box">
          <div className="roller-needle" />
          <table>
            <tbody>
              <tr
                className="roller-movable"
                style={{
                  left: scrollSize,
                  transitionDuration: `${props.duration}s`,
                }}
              >
                {users.map(item => (
                  <td>
                    <div className="roller-cell">
                      <img src={item.imageUrl} alt="logo" />
                      <span className="roller-label">{item.title}</span>
                    </div>
                  </td>
                ))}
              </tr>
            </tbody>
          </table>
        </div>
        <span
          className="raffle-winner"
          style={{ animationDelay: `${props.duration + 0.1}s` }}
        >
          {winner !== null && winner.title}
        </span>
      </div>
    </div>
  );
};

CSGORaffle.propTypes = {
  duration: PropTypes.number,
  onClose: PropTypes.func.isRequired,
  onWin: PropTypes.func.isRequired,
};
CSGORaffle.defaultProps = {
  duration: 7,
};

export default CSGORaffle;
