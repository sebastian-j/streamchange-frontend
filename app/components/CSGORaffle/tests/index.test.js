import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import CSGORaffle from '../index';

const renderer = new ShallowRenderer();

describe('<CSGORaffle />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(<CSGORaffle onClose={() => 0} onWin={() => 0} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
