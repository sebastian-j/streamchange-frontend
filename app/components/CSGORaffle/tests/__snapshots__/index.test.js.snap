// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CSGORaffle /> should render and match the snapshot 1`] = `
<div
  className="dialog-root"
>
  <button
    className="dialog-backdrop"
    onClick={[Function]}
    type="button"
  />
  <div
    className="raffle-dialog"
  >
    <div
      className="roller-box"
    >
      <div
        className="roller-needle"
      />
      <table>
        <tbody>
          <tr
            className="roller-movable"
            style={
              Object {
                "left": 0,
                "transitionDuration": "7s",
              }
            }
          />
        </tbody>
      </table>
    </div>
    <span
      className="raffle-winner"
      style={
        Object {
          "animationDelay": "7.1s",
        }
      }
    />
  </div>
</div>
`;
